src/producer.rs

```diff
@@ -1,242 +1,343 @@
+#[derive(Debug)]
+pub enum ArchiveType {
+    Zip(RefCell<ZipArchive<File>>),
+    Dir(PathBuf),
 }
 
-fn open_archive(path: &str) -> ZipArchive<File> {
-    let file = File::open(&path).expect(format!("Failed to open ZIP file '{}'.", path).as_str());
-    ZipArchive::new(file).expect(format!("Failed to parse ZIP file: {}", path).as_str())
+#[derive(Debug)]
+pub struct Archive {
+    pub name: String,
+    pub item: RefCell<ArchiveType>,
 }
 
-fn extract_file(zip_file: &mut zip::read::ZipFile, path: &PathBuf) {
-    let mut file = File::create(&path).expect("Failed to create file");
-    io::copy(zip_file, &mut file).expect("Failed to copy file from ZIP");
-}
+impl Archive {
 
-fn zip_producer(tmp_dir: &Path, zip_files: &[&String], queue: &WorkQueue, ignore_orphan_gcno: bool, is_llvm: bool) -> Option<Vec<u8>> {
-    let mut gcno_archive: Option<ZipArchive<File>> = None;
-    let mut gcda_archives: Vec<(&String,ZipArchive<File>)> = Vec::new();
-    let mut info_archives: Vec<(&String,ZipArchive<File>)> = Vec::new();
-
-    let mut path_mapping_file = None;
-
-    for zip_file in zip_files.iter() {
-        let archive = open_archive(zip_file);
-        if zip_file.contains("gcno") {
-            gcno_archive = Some(archive);
-        } else if zip_file.contains("gcda") {
-            gcda_archives.push((zip_file, archive));
-        } else if zip_file.contains("info") || zip_file.contains("grcov") || zip_file.contains("jsvm") {
-            info_archives.push((zip_file, archive));
+    fn insert_vec<'a>(&'a self, filename: String, map: &RefCell<HashMap<String, Vec<&'a Archive>>>) {
+        let mut map = map.borrow_mut();
+        if map.contains_key(&filename) {
+            let vec = map.get_mut(&filename).unwrap();
+            vec.push(self);
         } else {
-            panic!("Unsupported archive type.");
+            let mut vec = Vec::new();
+            vec.push(self);
+            map.insert(filename, vec);
         }
     }
 
-    if gcno_archive.is_some() {
-        assert!(!gcda_archives.is_empty());
+    pub fn get_name(&self) -> &String {
+        &self.name
     }
-    if !gcda_archives.is_empty() {
-        assert!(gcno_archive.is_some());
+
+    pub fn explore<'a>(&'a mut self,
+                       gcnos: &RefCell<HashMap<String, &'a Archive>>,
+                       gcdas: &RefCell<HashMap<String, Vec<&'a Archive>>>,
+                       infos: &RefCell<HashMap<String, Vec<&'a Archive>>>,
+                       linkeds: &RefCell<HashMap<String, &'a Archive>>) {
+        match *self.item.borrow() {
+            ArchiveType::Zip(ref zip) => {
+                let mut zip = zip.borrow_mut();
+                for i in 0..zip.len() {
+                    let filename = zip.by_index(i).unwrap();
+                    let filename = filename.name();
+                    let path = PathBuf::from(filename);
+                    match path.extension() {
+                        Some(ext) => match ext.to_str().unwrap() {
```

Comment:
It's a pity that we have to replicate this code twice, can we share it somehow between the two branches?